src/services/download/http.rs

```diff
@@ -43,85 +43,83 @@ fn join_url_encoded(base: &Url, path: &SourceLocation) -> Result<Url, ()> {
+        for (key, value) in source.headers.iter() {
+            if let Ok(key) = header::HeaderName::from_bytes(key.as_bytes()) {
+                builder.header(key, value.as_str());
+            }
+        }
+        builder.header(header::USER_AGENT, USER_AGENT);
+        // This timeout is for the entire HTTP download *including* the response stream
+        // itself, in contrast to what the Actix-Web docs say. We have tested this
+        // manually.
+        //
+        // The intent is to disable the timeout entirely, but there is no API for that.
+        builder.timeout(Duration::from_secs(9999));
+        builder.finish()
+    })
+    .await
+}
+
+pub async fn download_source(
     source: Arc<HttpSourceConfig>,
     download_path: SourceLocation,
     destination: PathBuf,
-) -> Box<dyn Future<Item = DownloadStatus, Error = DownloadError>> {
+) -> Result<DownloadStatus, DownloadError> {
     // This can effectively never error since the URL is always validated to be a base URL.
     // Though unfortunately this happens outside of this service, so ideally we'd fix this.
     let download_url = match join_url_encoded(&source.url, &download_path) {
         Ok(x) => x,
-        Err(_) => return Box::new(future::ok(DownloadStatus::NotFound)),
+        Err(_) => return Ok(DownloadStatus::NotFound),
     };
     log::debug!("Fetching debug file from {}", download_url);
-    let response = clone!(download_url, source, || {
-        http::follow_redirects(
-            download_url.clone(),
-            MAX_HTTP_REDIRECTS,
-            clone!(source, |url| {
-                let mut builder = client::get(url);
-
-                for (key, value) in source.headers.iter() {
-                    if let Ok(key) = header::HeaderName::from_bytes(key.as_bytes()) {
-                        builder.header(key, value.as_str());
-                    }
-                }
-
-                builder.header(header::USER_AGENT, USER_AGENT);
-
-                // This timeout is for the entire HTTP download *including* the response stream
-                // itself, in contrast to what the Actix-Web docs say. We have tested this
-                // manually.
-                //
-                // The intent is to disable the timeout entirely, but there is no API for that.
-                builder.timeout(Duration::from_secs(9999));
-                builder.finish()
-            }),
-        )
-    });
-
-    let response = Retry::spawn(
-        ExponentialBackoff::from_millis(10).map(jitter).take(3),
-        response,
-    );
-
-    let response = response.map_err(|e| match e {
-        tokio_retry::Error::OperationError(e) => e,
-        e => panic!("{}", e),
-    });
-
-    let response = response.then(move |result| match result {
+
+    let mut backoff = ExponentialBackoff::from_millis(10).map(jitter).take(3);
+    let response = loop {
+        let result = start_request(&source, download_url.clone()).await;
```

Comment:
just wondering, since `start_request` does the `follow_redirects`.
Lets suppose we have redirects of *A -> B -> C*, and C is down, that means we re-do the redirects a couple of times.
I think those cases are quite rare and not worth optimizing for, just to keep this in mind.